### PR TITLE
Add Hadoop metrics receiver

### PR DESCRIPTION
**Pull request by Mrod1598**
Add hadoop metrics receiver

### CODE REVIEW

File: confgenerator/testdata/invalid/linux/metrics-receiver_hadoop_malformed_url/input.yaml
Comment (qingling128) [inline confgenerator/testdata/invalid/linux/metrics-receiver_hadoop_malformed_url/input.yaml:17]: It's preferred for the receiver and pipeline IDs to stay the same as the receiver type in the sample configuration. E.g.  "s/{APP}_metrics/{APP}", "s/{APP}_pipeline/{APP}"

File: confgenerator/testdata/valid/linux/metrics-receiver_hadoop/input.yaml
Comment (qingling128) [inline confgenerator/testdata/valid/linux/metrics-receiver_hadoop/input.yaml:22]: We could remove `collect_interval` from here as it's optional.

For the context: There should at least be one valid unit test that verifies the absolute minimum user configuration. All optional configurations should be absent for this case including "collection_interval".

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:11]: Chances are a bit higher that users will use the default package manager (e.g. yum, apt-get, zyypper) to install a package instead of a pinned version of the application.

Can we adjust it to do something like `yum install hadoop` instead? Similar requests for the other `install` scripts.

File: apps/hadoop.go
Comment (qingling128) [inline apps/hadoop.go:48]: Can we refactor this part a bit to use the common util by https://github.com/GoogleCloudPlatform/ops-agent/pull/370?

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (Mrod1598) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:11]: there doesn't seem to be an available package for hadoop.

File: confgenerator/testdata/valid/linux/metrics-receiver_hadoop/input.yaml
Comment (Mrod1598) [reply confgenerator/testdata/valid/linux/metrics-receiver_hadoop/input.yaml:22]: Updated

File: apps/hadoop.go
Comment (Mrod1598) [reply apps/hadoop.go]: Updated

File: confgenerator/testdata/invalid/linux/metrics-receiver_hadoop_malformed_url/input.yaml
Comment (Mrod1598) [reply confgenerator/testdata/invalid/linux/metrics-receiver_hadoop_malformed_url/input.yaml]: Updated

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:11]: I agree, I don't see any packaging for Hadoop. This seems like a fairly common thing for Java apps - hopefully we can get away from manual installation when we move on to non-Java apps.

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:16]: Use an `EnvironmentFile` in the systemd unit so this isn't duplicated so many times?

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (qingling128) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:11]: Ah NVM for this app then.

File: integration_test/third_party_apps_data/agent/ops-agent/linux/enable_hadoop
Comment (qingling128) [inline integration_test/third_party_apps_data/agent/ops-agent/linux/enable_hadoop:25]: Remove `collection_interval` here as well?

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:138]: ... I don't want to know why the hadoop user needs to be able to ssh to itself, do I?

File: integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install:6]: These scripts are nearly identical, can we refactor to share common code?

**Review by qingling128 — APPROVED** [review]
LGTM modulo the remaining open threads are addressed.

File: integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install
Comment (Mrod1598) [reply integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install:6]: So, what is actually copied into the environment? is it the whole hadoop folder?

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (Mrod1598) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:138]: I promise it's a [requirement](https://phoenixnap.com/kb/install-hadoop-ubuntu).

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (Mrod1598) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:16]: We can't use an environment file because the environment variables are used in three spots `systemd unit`, `core-site.xml`, and `hadoop users bashrc` and the environement file, I don't believe, allows the export for the environment file.

File: integration_test/third_party_apps_data/agent/ops-agent/linux/enable_hadoop
Comment (Mrod1598) [reply integration_test/third_party_apps_data/agent/ops-agent/linux/enable_hadoop]: Removed

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:138]: Aha, I found it: https://stackoverflow.com/a/20477333
Apparently it uses ssh to start the daemons on data nodes. I'm kinda surprised that is needed on a single-node cluster.

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:16]: I think you could put the export statement on a separate line e.g.
```
A=1
B=2
export A B
```
and systemd would ignore the export line. I haven't tested it though.

File: integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install:6]: I don't know the answer to that :/
@martijnvans perhaps?

Comment (qingling128) [reply integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install:6]: I was thinking about something similar. It would likely require some refactoring of the testing infra though to support a shared snippet.

```
- hadoop
   - install
      - common
      - debian_ubuntu
      - centos_rhel
      - sles
```

A follow up PR for this refactoring is better IMO instead of trying to doing that in this PR. So that this PR gets unblocked first, and we could keep the refactoring separate from a new feature.

Comment (martijnvans) [reply integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install:6]: ling is right, we'd need to introduce some kind of "common" script that gets run for all platforms. Such a change should be pretty easy.

**Comment by qingling128** [comment]
Integration tests seem to be failing for Hadoop:

image

File: integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install
Comment (qingling128) [reply integration_test/third_party_apps_data/applications/hadoop/debian_ubuntu/install:6]: Cool. I'll leave this thread unresolved for easy reference (GitHub UI makes it hard to find resolved comments). It does not block this PR from being merged.

File: integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install
Comment (Mrod1598) [reply integration_test/third_party_apps_data/applications/hadoop/centos_rhel/install:16]: Haven't been able to apply this successfully. I can't seem to find anything online as an example of anyone doing this either. Seems that once we add in the ability to refactor out this into a separate common script this will be a lot cleaner anyways as a side effect, if that was your concern.

File: apps/hadoop.go
Comment (qingling128) [inline apps/hadoop.go:29]: Now that the Elasticsearch PR is merged, can we make use of the common util here as well? https://github.com/GoogleCloudPlatform/ops-agent/blame/c145e38a9640f55eb46761398ad5446ecc9cea07/confgenerator/config.go#L479

Since there is a planned follow up PR for Hadoop, and this change is not user visible, I'm okay with doing it in the follow up PR as well.

**Review by qingling128 — APPROVED** [review]
LGTM

File: apps/hadoop.go
Comment (Mrod1598) [reply apps/hadoop.go]: Updated!

Comment (qingling128) [reply apps/hadoop.go]: Thanks!

**Review by qingling128 — APPROVED** [review]
LGTM. Waiting for the integration tests before merging

**Comment by qingling128** [comment]
`third party apps integration test (Debian)` is failing for debian 10 on tomcat, which is a known issue that is being fixed by a separate PR https://github.com/GoogleCloudPlatform/ops-agent/pull/344.

`third party apps integration test (CentOS)` passed in few other PRs but failed for this one. However the error seems to be a service account key issue. Might not be related to this PR.

File: docs/hadoop.md
Comment (qingling128) [inline docs/hadoop.md:4]: Just realized that Hadoop has some pending discussions going on in the `Category 3 - Regional metrics and master metrics` section of the `Duplicate / Overlapping Metrics Issue from Clustered Applications` doc.

We'll need to resolve that conversion before releasing this.

**Review by qingling128 — CHANGES_REQUESTED** [review]

**Comment by Mrod1598** [comment]
> `third party apps integration test (Debian)` is failing for debian 10 on tomcat, which is a known issue that is being fixed by a separate PR #344.
> 
> `third party apps integration test (CentOS)` passed in few other PRs but failed for this one. However the error seems to be a service account key issue. Might not be related to this PR.


It did seem to be a one off issue.
image

File: docs/hadoop.md
Comment (qingling128) [reply docs/hadoop.md:4]: Per the recent decision in the Duplicate / Overlapping Metrics Issue from Clustered Applications, let's add some context here:

> It collects metrics specific to the local region server, as well as metrics presented by the Master node if the node being monitored is indeed the Master. For High Availability configurations, it is recommended for every master node to report cluster metrics, even though they are identical, to avoid single point of failures when one master goes down.

**Review by qingling128 — APPROVED** [review]
LGTM modulo 2 minor comments.

File: apps/hadoop.go
Comment (Mrod1598) [reply apps/hadoop.go]: Added

File: docs/hadoop.md
Comment (Mrod1598) [reply docs/hadoop.md]: Added

**Review by qingling128 — APPROVED** [review]
LGTM

**Pull request merged**
